### PR TITLE
Initiate non blocking threads for collection

### PR DESCRIPTION
**Pull request by SunnySrivastava1984**
Instantiate threads to collect FRU in a separate thread so that main thread is not blocked till the collection is in progress.

### CODE REVIEW

File: src/manager.cpp [hunk 1]
```diff
@@ -111,12 +114,62 @@ void Manager::SetTimerToDetectSVPDOnDbus()
...
             m_worker->collectFrusFromJson();
+        }
+    });
+}
+
+void Manager::SetTimerToDetectVpdCollectionStatus()
+{
+    static constexpr auto MAX_RETRY = 5;
+
+    static boost::asio::steady_timer timer(*m_ioContext);
```
Comment (branupama) [inline src/manager.cpp:130]: local variables should starts with l_

File: src/manager.cpp [hunk 2]
```diff
@@ -111,12 +114,62 @@ void Manager::SetTimerToDetectSVPDOnDbus()
...
+        {
+            if (timerCount == MAX_RETRY)
+            {
+                timer.cancel();
+                logging::logMessage("Taking too long. Some logic needed here?");
+            }
+            else
+            {
+                logging::logMessage(
+                    "Waiting... FRU VPD collection in progress");
```
Comment (branupama) [inline src/manager.cpp:169]: is in progress .?

File: src/manager.cpp [hunk 3]
```diff
@@ -111,12 +114,62 @@ void Manager::SetTimerToDetectSVPDOnDbus()
...
+        if (m_worker->isAllFruCollectionDone())
+        {
+            // cancel the timer
+            timer.cancel();
             m_interface->set_property("CollectionStatus",
                                       std::string("Completed"));
         }
+        else
+        {
+            if (timerCount == MAX_RETRY)
```
Comment (branupama) [inline src/manager.cpp:161]: timerCount should increment ?

File: include/worker.hpp [hunk 4]
```diff
@@ -356,5 +366,14 @@ class Worker
 
     // Path to config JSON if applicable.
     std::string& m_configJsonPath;
+
+    // Keeps track of active thread doing VPD collection.
```
Comment (branupama) [inline include/worker.hpp:370]: Nits: threads count

File: src/worker.cpp [hunk 5]
```diff
@@ -1173,24 +1170,36 @@ void Worker::collectFrusFromJson()
 
         logging::logMessage("Parsing triggered for FRU = " + vpdFilePath);
 
-        listOfFutureObject.push_back(
-            std::async(&Worker::parseAndPublishVPD, this, vpdFilePath));
-    }
+        std::thread{[&vpdFilePath, this]() {
+            auto futureObject = std::async(&Worker::parseAndPublishVPD, this,
```
Comment (branupama) [inline src/worker.cpp:1174]: l_futureObject

File: src/worker.cpp [hunk 6]
```diff
@@ -1173,24 +1170,36 @@ void Worker::collectFrusFromJson()
...
+                                           vpdFilePath);
+            activeCollectionThreadCount++;
 
-    for (auto& aFutureObj : listOfFutureObject)
-    {
-        std::tuple<bool, std::string> threadInfo = aFutureObj.get();
+            std::tuple<bool, std::string> threadInfo = futureObject.get();
+            if (!std::get<0>(threadInfo))
+            {
+                activeCollectionThreadCount--;
```
Comment (branupama) [inline src/worker.cpp:1181]: can we keep this outside ?

File: src/manager.cpp (hunk 1 above)
Comment (SunnySrivastava1984) [reply src/manager.cpp]: done

File: src/manager.cpp (hunk 3 above)
Comment (SunnySrivastava1984) [reply src/manager.cpp]: Fixed.

File: src/worker.cpp (hunk 6 above)
Comment (SunnySrivastava1984) [reply src/worker.cpp]: yes we can.

File: src/manager.cpp [hunk 7]
```diff
@@ -111,12 +114,63 @@ void Manager::SetTimerToDetectSVPDOnDbus()
...
+            // cancel the timer
+            l_timer.cancel();
             m_interface->set_property("CollectionStatus",
                                       std::string("Completed"));
         }
+        else
+        {
+            if (l_timerRetry == MAX_RETRY)
+            {
+                timer.cancel();
```
Comment (branupama) [inline src/manager.cpp:163]: l_timer ?

File: src/worker.cpp [hunk 8]
```diff
@@ -1173,24 +1170,39 @@ void Worker::collectFrusFromJson()
...
-        }
+            // thread returned.
+            m_activeCollectionThreadCount--;
+
+            if (!std::get<0>(l_threadInfo))
+            {
+                logging::logMessage("Processing failed for = " +
+                                    std::get<1>(l_threadInfo));
+            }
+            else if (std::get<0>(l_threadInfo))
```
Comment (branupama) [inline src/worker.cpp:1189]: can we use else here directly. ?

**Review by branupama — CHANGES_REQUESTED** [review]

**Review by branupama — APPROVED** [review]

**Pull request merged**